src/main/java/frc/trigon/robot/subsystems/arm/Arm.java

```diff
@@ -0,0 +1,102 @@
+        armIO.updateInputs(armInputs);
+        Logger.processInputs("Arm", armInputs);
+        updateMechanism();
+    }
+
+    boolean isAtAngle(Rotation2d targetAngle) {
+        return Math.abs(armInputs.anglePositionDegrees - targetAngle.getDegrees()) <= ArmConstants.TOLERANCE;
+    }
+
+    boolean isAtPosition(double targetElevatorPosition) {
+        return Math.abs(armInputs.elevatorPositionRevolution - targetElevatorPosition) <= ArmConstants.TOLERANCE;
+    }
+
+    boolean isElevatorOpening(double targetElevatorPosition) {
+        return armInputs.elevatorPositionRevolution < targetElevatorPosition;
+    }
+
+    void setTargetAngleFromProfile() {
+        if (angleMotorProfile == null) {
+            armIO.stopAngleMotors();
+            return;
+        }
+
+        TrapezoidProfile.State targetState = angleMotorProfile.calculate(getAngleMotorProfileTime());
+        ArmConstants.TARGET_POSITION_LIGAMENT.setAngle(targetState.position);
+        armIO.setTargetAngleState(targetState);
+    }
+
+    void setTargetElevatorPositionFromProfile() {
+        if (elevatorMotorProfile == null) {
+            armIO.stopElevatorMotors();
+            return;
+        }
+
+        TrapezoidProfile.State targetState = elevatorMotorProfile.calculate(getElevatorMotorProfileTime() + ArmConstants.RETRACTED_ARM_LENGTH);
+        ArmConstants.TARGET_POSITION_LIGAMENT.setLength(targetState.position);
+        armIO.setTargetElevatorState(targetState);
+    }
+
+    void generateAngleMotorProfile(Rotation2d targetAngle, double speedPercentage) {
+        angleMotorProfile = new TrapezoidProfile(
+                Conversions.scaleConstraints(ArmConstants.ANGLE_CONSTRAINTS, speedPercentage),
+                new TrapezoidProfile.State(targetAngle.getDegrees(), 0),
+                new TrapezoidProfile.State(armInputs.anglePositionDegrees, armInputs.angleVelocityDegreesPerSecond)
+        );
+
+        lastAngleMotorProfileGenerationTime = Timer.getFPGATimestamp();
+    }
+
+    void generateElevatorMotorProfile(double targetElevatorPosition, double speedPercentage) {
+        elevatorMotorProfile = new TrapezoidProfile(
+                Conversions.scaleConstraints(ArmConstants.ELEVATOR_CONSTRAINTS, speedPercentage),
+                new TrapezoidProfile.State(targetElevatorPosition, 0),
+                new TrapezoidProfile.State(armInputs.elevatorPositionRevolution, armInputs.elevatorVelocityRevolutionsPerSecond)
+        );
+
+        lastElevatorMotorProfileGenerationTime = Timer.getFPGATimestamp();
+    }
+
+    private double getAngleMotorProfileTime() {
+        return Timer.getFPGATimestamp() - lastAngleMotorProfileGenerationTime;
+    }
+
+    private double getElevatorMotorProfileTime() {
+        return Timer.getFPGATimestamp() - lastElevatorMotorProfileGenerationTime;
+    }
+
+    private void updateMechanism() {
+        ArmConstants.ARM_LIGAMENT.setLength(armInputs.elevatorPositionRevolution + ArmConstants.RETRACTED_ARM_LENGTH);
```

Comment:
the retracted arm length is in meters, while the elevator position is in revolutions